src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValuesIterator.java

```diff
@@ -0,0 +1,157 @@
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * An abstract class that prvides an iterator to iterate over KNNVectors, as KNNVectors are stored as different
+ * representation like {@link BinaryDocValues}, {@link FloatVectorValues}, FieldWriter etc.
+ */
+public interface KNNVectorValuesIterator {
+
+    /**
+     * Returns the current docId where the iterator is pointing to.
+     * @return int
+     */
+    int docId();
+
+    /**
+     * Advances to a specific docId. Ensure that the passed docId is greater than current docId where Iterator is
+     * pointing to, otherwise
+     * {@link IOException} will be thrown
+     * @return int
+     * @throws IOException if we are not able to move to the passed docId.
+     */
+    int advance(int docId) throws IOException;
+
+    /**
+     * Move to nextDocId. If no more docs are present then {@link DocIdSetIterator#NO_MORE_DOCS} will be returned.
+     * @return int
+     * @throws IOException if we cannot move to next docId
+     */
+    int nextDoc() throws IOException;
+
+    /**
+     * Return a {@link DocIdSetIterator}
+     * @return {@link DocIdSetIterator}
+     */
+    DocIdSetIterator getDocIdSetIterator();
+
+    /**
+     * Total number of live doc which will the iterator will iterate upon.
+     * @return long: total number of live docs
+     */
+    long liveDocs();
+
+    /**
+     * A DocIdsIteratorValues provides a common iteration logic for all Values that implements
+     * {@link DocIdSetIterator} interface. Example: {@link BinaryDocValues}, {@link FloatVectorValues} etc.
+     */
+    class DocIdsIteratorValues implements KNNVectorValuesIterator {
+        protected DocIdSetIterator docIdSetIterator;
+
+        DocIdsIteratorValues(@NonNull final DocIdSetIterator docIdSetIterator) {
+            this.docIdSetIterator = docIdSetIterator;
+        }
+
+        @Override
+        public int docId() {
+            return docIdSetIterator.docID();
+        }
+
+        @Override
+        public int advance(int docId) throws IOException {
+            return docIdSetIterator.advance(docId);
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return docIdSetIterator.nextDoc();
+        }
+
+        @Override
+        public DocIdSetIterator getDocIdSetIterator() {
+            return docIdSetIterator;
+        }
+
+        @Override
+        public long liveDocs() {
+            if (docIdSetIterator instanceof BinaryDocValues) {
+                return KNNCodecUtil.getTotalLiveDocsCount((BinaryDocValues) docIdSetIterator);
+            } else if (docIdSetIterator instanceof FloatVectorValues || docIdSetIterator instanceof ByteVectorValues) {
+                return docIdSetIterator.cost();
+            }
+            throw new IllegalArgumentException(
+                "DocIdSetIterator present is not of valid type. Valid types are: BinaryDocValues, FloatVectorValues and ByteVectorValues"
```

Comment:
I think the validation should be in the constructor. This might force us to validate in other methods or logic being added in this class. The earlier the validation the better. Let me know what you think

Reply:
I agree that validation can be moved to the constructor, but even in that case this exception will be present that is curse of `if/else`. we need to have a else condition. Lets say even if I move this logic via Map<itr, functional interface> I would still need to throw the exception there as the output of map can be null for an iterator which is not present in map.

Reply:
Added the validations in constructor too.